generators/Reporting.fs

```diff
@@ -0,0 +1,92 @@
+    | Custom of string
+    | Unimplemented of string
+    | MissingData of string
+    | Deprecated of string
+    | UpToDate of string
+    | Outdated of string * string * string
+    
+let nameFromSummaryType = function
+    | Custom n -> n
+    | Unimplemented  n -> n
+    | MissingData n -> n
+    | Deprecated n -> n
+    | UpToDate n -> n
+    | Outdated (n,_,_) -> n
+    
+let private summarizeExercise options (parseCanonicalData':string -> CanonicalData) (exercise:Exercise)  =
+
+    match exercise with
+    | Exercise.Custom custom ->                 SummaryTypes.Custom custom.Name
+    | Exercise.Unimplemented unimplemented ->   SummaryTypes.Unimplemented unimplemented.Name
+    | Exercise.MissingData missingData ->       SummaryTypes.MissingData missingData.Name
+    | Exercise.Deprecated deprecated ->         SummaryTypes.Deprecated deprecated.Name
+    | Exercise.Generator generator ->
+        let cData = parseCanonicalData' generator.Name
+        
+        match generator.ReadVersion (), cData.Version with
+        | a,b when a.Equals b ->  UpToDate generator.Name
+        | a,b -> Outdated (generator.Name,a,b)
+        
+let listExercises options =
+    Log.Information(sprintf "Listing exercises with status %s..." (options.Status.Value.ToString()))
+    
+    let parseCanonicalData' = CanonicalData.parseCanonicalData options
+    
+    let allExercises = createExercises options
+    
+    let longestNameLength = allExercises |> List.map exerciseName |> List.map (fun e -> e.Length) |> List.max
+    let indentAfter (name:string) = String.replicate (longestNameLength+2-name.Length) " "
+    
+    let exerciseGroups = allExercises
+                            |> List.filter (filterByStatus options parseCanonicalData')
+                            |> List.map (summarizeExercise options parseCanonicalData')
+                            |> List.groupBy (fun x -> x.GetType())
+    
+    exerciseGroups |> List.iter (fun (groupType,group) ->
+        let description = match group.Head with
+                          | SummaryTypes.Custom _ -> "have customized tests"
+                          | SummaryTypes.Unimplemented _ -> "are missing test generator"
+                          | SummaryTypes.MissingData _ -> "are missing canonical data"
+                          | SummaryTypes.Deprecated _ -> "are deprecated"
+                          | SummaryTypes.UpToDate _ -> "are up to date"
+                          | SummaryTypes.Outdated _ -> "are outdated"
+                          
+        Log.Information("{num} exercises {description}:", group.Length, description)
+        
+        group |> List.iter (fun exercise -> 
+            match exercise with
+            | SummaryTypes.Outdated (name,oldVer,newVer) -> Log.Information(" {name}{indent}({oldVer} -> {newVer})", name, indentAfter name, oldVer, newVer)
+            | _ -> Log.Information(" {name}", nameFromSummaryType exercise)
+        )
+            
+        printfn ""
+        
```

Comment:
Could you remove this unneeded empty line?

Reply:
This white space was deliberate, as well, to separate the groups when you print them all:
```
[10:45:16 INF] 4 exercises are deprecated:
[10:45:16 INF]    binary
[10:45:16 INF]    hexadecimal
[10:45:16 INF]    octal
[10:45:16 INF]    trinary

[10:45:16 INF] 1 exercises have customized tests:
[10:45:16 INF]    diamond

[10:45:16 INF] 3 exercises are missing test generator:
[10:45:16 INF]    grade-school
[10:45:16 INF]    pythagorean-triplet
[10:45:16 INF]    sgf-parsing
```

(but if you prefer, I'll remove it)